packages/run-protocol/src/econ-behaviors.js

```diff
@@ -167,6 +167,7 @@ export const setupAmm = async ({
     E(instanceAdmin).update('ammGovernor', g.instance),
   ]);
 };
+harden(setupAmm);
```

Comment:
I'm not surprised to see lots of functions exported but not hardened. Fortunately it is not very dangerous. But it is a hazard we need to avoid. It leaves mutable top-level state, such that two parties importing the same function from the same module instance can now communicate without calling the function. This isn't POLA and breaks reasoning by separation.

Reply:
> it is a hazard we need to avoid

Could missing `harden` be detected automatically? i.e. are there consistent rules for when to use harden or not?

Reply:
Yes, there are consistent rules:
> all objects made by literal expressions (object literals, array literals, the many forms of function literals) must be tamper-proofed with harden before they can be aliased or escape from their static context of origin -- https://github.com/endojs/Jessie#must-freeze-api-surface-before-use

Automatic detection is an open research problem, IIUC. https://github.com/agoric-labs/jessica/issues/35

Reply:
- a) One abstract rule is that an object should not escape the abstraction it is born in without first being hardened.
- b) A related rule is that module instances should be pure --- they should have no mutable top level state.

On #b, We have tried several times to write down static rules that would enforce these abstract rules. I think @michaelfig has grammatical restrictions for pure Jessie modules somewhere. @michaelfig ?

On #a, a conservative rule that works almost all the time is to just put a `harden(...)` or `Far(..., ...)` around the outer literal (object literal, array literal, function literal, class literal, regexp literal) where the object is first born. Then, no one gets to see the object in its pre-hardened state. 

Occasionally, we do have valid code where we locally mutate the object before hardening it, but still harden it before it might escape. To enforce that statically would involve treating all variables holding the object as "owning" variables in the reference-capability sense. Or as linear or affine variables or something. The key is that we statically know that the object hasn't escaped via an alias before it's been hardened. We have sketched static rules for this but never got very far. Fortunately, the need for such delayed hardening happens even more rarely than I expected. We can almost always simply follow the conservative harden-the-literal rule above.

Reply:
On #b, a conservative rule is to only have top level const declarations of hardened things. Again, this conservative rule works almost always.

For exported functions, we follow a pattern of a slightly delayed harden:
```js
export const foo = x => x * x;
harden(foo);
```
This still hardens `foo` before anything could observe `foo` in its unhardened state.

Reply:
Thanks @dckc . I answered before I saw your response.

Reply:
Would an imperfect but worthwhile rule be that named exports from a module have a corresponding `harden`? We could pretty easily lint for that.

E..g,
**FAIL**
```js
export const sayHi = () => {
  console.log("hi");
};
```
**PASS**
```js
export const sayHi = () => {
  console.log("hi");
};
harden(sayHi);

// eslint-disable-next-line endo/hardened-exports
export const foo = {};
```

Reply:
Yes!

Reply:
You may need to accept `harden` or `Far` there for the rule.